File: offchain/midgard-sdk/src/types/contracts/active-operators.ts
import { Data } from "@lucid-evolution/lucid";
import { POSIXTimeSchema } from "./common.js";

export const DatumSchema = Data.Object({
  commitmentTime: Data.Nullable(POSIXTimeSchema),
});
export type Datum = Data.Static<typeof DatumSchema>;
export const Datum = DatumSchema as unknown as Datum;

export const SpendRedeemerSchema = Data.Enum([
  Data.Literal("ListStateTransition"),
  Data.Object({
    UpdateCommitmentTime: Data.Object({
      activeNodeInputIndex: Data.Integer(),
      prevActiveNodeRefInputIndex: Data.Integer(),
      hubOracleRefInputIndex: Data.Integer(),
      stateQueueRedeemerIndex: Data.Integer(),
    }),
  }),
]);
export type SpendRedeemer = Data.Static<typeof SpendRedeemerSchema>;
export const SpendRedeemer = SpendRedeemerSchema as unknown as SpendRedeemer;

export const MintRedeemerSchema = Data.Enum([
  Data.Literal("Init"),
  Data.Literal("Deinit"),
  Data.Object({
    ActivateOperator: Data.Object({
      newActiveOperatorKey: Data.Bytes(),
      hubOracleRefInputIndex: Data.Integer(),
      activeOperatorAppendedNodeOutputIndex: Data.Integer(),
      activeOperatorAnchorNodeOutputIndex: Data.Integer(),
      registeredOperatorsRedeemerIndex: Data.Integer(),
    }),
  }),
  Data.Object({
    RemoveOperatorSlashBond: Data.Object({
      slashedActiveOperatorKey: Data.Bytes(),
      hubOracleRefInputIndex: Data.Integer(),
      activeOperatorSlashedNodeInputIndex: Data.Integer(),
      activeOperatorAnchorNodeInputIndex: Data.Integer(),
      stateQueueRedeemerIndex: Data.Integer(),
    }),
  }),
  Data.Object({
    RetireOperator: Data.Object({
      activeOperatorKey: Data.Bytes(),
      hubOracleRefInputIndex: Data.Integer(),
      activeOperatorRemovedNodeInputIndex: Data.Integer(),
      activeOperatorAnchorNodeInputIndex: Data.Integer(),
      retiredOperatorInsertedNodeOutputIndex: Data.Integer(),
      retiredOperatorsRedeemerIndex: Data.Integer(),
    }),
  }),
]);
export type MintRedeemer = Data.Static<typeof MintRedeemerSchema>;
export const MintRedeemer = MintRedeemerSchema as unknown as MintRedeemer;

File: offchain/src/Midgard/Types/ActiveOperators.hs
{-# LANGUAGE TemplateHaskell #-}

module Midgard.Types.ActiveOperators (
  NodeData (..),
  Datum,
  SpendRedeemer (..),
  MintRedeemer (..),
  rootAssetName,
  nodeAssetNamePrefix,
  nodeAssetNamePrefixLen,
) where

import Data.ByteString (ByteString)
import Data.ByteString.Char8 qualified as BS8
import GHC.Generics (Generic)

import Cardano.Api qualified as C
import PlutusLedgerApi.V3 (BuiltinByteString, POSIXTime, PubKeyHash)
import PlutusTx.Blueprint (HasBlueprintDefinition, definitionRef)
import PlutusTx.Blueprint.TH (makeIsDataSchemaIndexed)

import Ply (PlyArg)

import Midgard.Types.LinkedList qualified as LinkedList

rootAssetName :: C.AssetName
rootAssetName = C.UnsafeAssetName $ BS8.pack "MIDGARD_ACTIVE_OPERATORS"

nodeAssetNamePrefix :: ByteString
nodeAssetNamePrefix = BS8.pack "MACT"

nodeAssetNamePrefixLen :: Int
nodeAssetNamePrefixLen = BS8.length nodeAssetNamePrefix

newtype NodeData = NodeData
  { bondUnlockTime :: Maybe POSIXTime
  }
  deriving stock (Eq, Show, Generic)
  deriving anyclass (HasBlueprintDefinition)

$( makeIsDataSchemaIndexed
     ''NodeData
     [ ('NodeData, 0)
     ]
 )

type Datum = LinkedList.Element BuiltinByteString NodeData

data SpendRedeemer
  = ListStateTransition
  | UpdateBondHoldNewState
      { activeNodeInputIndex :: Integer
      , activeNodeOutputIndex :: Integer
      , hubOracleRefInputIndex :: Integer
      , stateQueueInputIndex :: Integer
      , stateQueueRedeemerIndex :: Integer
      }
  | UpdateBondHoldNewSettlement
      { activeNodeInputIndex :: Integer
      , activeNodeOutputIndex :: Integer
      , hubOracleRefInputIndex :: Integer
      , settlementInputIndex :: Integer
      , settlementRedeemerIndex :: Integer
      , newBondUnlockTime :: POSIXTime
      }
  deriving stock (Eq, Show, Generic)
  deriving anyclass (HasBlueprintDefinition)

$( makeIsDataSchemaIndexed
     ''SpendRedeemer
     [ ('ListStateTransition, 0)
     , ('UpdateBondHoldNewState, 1)
     , ('UpdateBondHoldNewSettlement, 2)
     ]
 )

instance PlyArg SpendRedeemer

-- Mint redeemer

data MintRedeemer
  = Init {outputIndex :: Integer}
  | Deinit {inputIndex :: Integer}
  | ActivateOperator
      { newActiveOperatorKey :: PubKeyHash
      , activeOperatorAnchorElementInputIndex :: Integer
      , activeOperatorAnchorElementOutputIndex :: Integer
      , activeOperatorInsertedNodeOutputIndex :: Integer
      , registeredOperatorsRedeemerIndex :: Integer
      }
  | RemoveOperatorBadState
      { slashedActiveOperatorKey :: PubKeyHash
      , hubOracleRefInputIndex :: Integer
      , activeOperatorAnchorElementInputIndex :: Integer
      , activeOperatorSlashedNodeInputIndex :: Integer
      , activeOperatorAnchorElementOutputIndex :: Integer
      , stateQueueRedeemerIndex :: Integer
      }
  | RemoveOperatorBadSettlement
      { slashedActiveOperatorKey :: PubKeyHash
      , hubOracleRefInputIndex :: Integer
      , activeOperatorAnchorElementInputIndex :: Integer
      , activeOperatorSlashedNodeInputIndex :: Integer
      , activeOperatorAnchorElementOutputIndex :: Integer
      , settlementInputIndex :: Integer
      , settlementRedeemerIndex :: Integer
      }
  | RetireOperator
      { activeOperatorKey :: PubKeyHash
      , activeOperatorAnchorElementInputIndex :: Integer
      , activeOperatorRemovedNodeInputIndex :: Integer
      , activeOperatorAnchorElementOutputIndex :: Integer
      , retiredOperatorsRedeemerIndex :: Integer
      }
  deriving stock (Eq, Show, Generic)
  deriving anyclass (HasBlueprintDefinition)

$( makeIsDataSchemaIndexed
     ''MintRedeemer
     [ ('Init, 0)
     , ('Deinit, 1)
     , ('ActivateOperator, 2)
     , ('RemoveOperatorBadState, 3)
     , ('RemoveOperatorBadSettlement, 4)
     , ('RetireOperator, 5)
     ]
 )

instance PlyArg MintRedeemer
